### PR TITLE
Introduce pre-commit hooks and GitHub Actions

### PR DESCRIPTION
**Pull request by yu-iskw**
## Motivation

pre-commit is very useful to apply static code analyzers including linters to code using git hooks. This pull request aims to implement basic pre-commit hooks to keep the code quality pr-agent great.

### CODE REVIEW

File: .github/workflows/pre-commit.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    branches: [main]
+
+jobs:
+  pre-commit:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-python@v5
+      # SEE https://github.com/pre-commit/action
+      - uses: pre-commit/action@v3.0.1
```
Comment (yu-iskw) [inline .github/workflows/pre-commit.yml:15]: This GitHub Action comes from https://github.com/pre-commit/action .

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+repos:
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v5.0.0
+    hooks:
+      - id: check-added-large-files
+      - id: check-toml
+      - id: check-yaml
+      - id: end-of-file-fixer
+      - id: trailing-whitespace
```
Comment (yu-iskw) [inline .pre-commit-config.yaml:15]: We can go with basic pre-commit hooks to format code. Then, we can take other hooks if ok.

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  # - repo: https://github.com/rhysd/actionlint
+  #   rev: v1.7.3
+  #   hooks:
+  #     - id: actionlint
+  - repo: https://github.com/pycqa/isort
+    # rev must match what's in dev-requirements.txt
+    rev: 5.13.2
+    hooks:
+      - id: isort
```
Comment (yu-iskw) [inline .pre-commit-config.yaml:24]: ditto

**Comment by mrT23** [comment]
Hi @yu-iskw , and thanks for the PR.

However, the world is divided into two - people who love linters, and people who hate them.
I personally really hate them. They mess up the new lines, and make the code less readable.

Its a matter of taste, but as the reop owner, this repo will follow my preferred taste.

So sorry, but it won't be merged at this point

**Comment by mrT23** [comment]
@yu-iskw thanks for the PR.

I was starting an answer describing how much I hate linters (especially the line breaks, which usually make the code unreadable), but when reviewing the results, I actually kind of like them. It mainly orders things in alphabetical order, which is good, and makes the spacing more consistent.

How will this hook work in practice ? what happens when I open a PR that is not complaint with its rules?

**Comment by yu-iskw** [comment]
@mrT23 

> How will this hook work in practice ? what happens when I open a PR that is not complaint with its rules?

The hooks enabled in the pull request automatically formats code. As long as pre-commit is installed locally, pre-commit hooks are automatically triggered when committing and pushing commits. 

Apart from formating code, some pre-commit hooks like bandit and semgrep helps us to automatically find vulnerabilities and security issues to some degree. That would be useful to keep our code secure. 

- https://github.com/PyCQA/bandit/blob/main/.pre-commit-hooks.yaml
- https://github.com/semgrep/pre-commit
- https://github.com/google/osv-scanner/blob/main/.pre-commit-hooks.yaml

**Review by mrT23 — APPROVED** [review]

**Comment by yu-iskw** [comment]
@mrT23 Thank you for merging!

**Pull request merged**